pkg/apis/configs/v1alpha1/configmappropagation_types.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ConfigMapPropagation is used to propagate configMaps from original namespace to current namespace
+type ConfigMapPropagation struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the ConfigMapPropagation
+	Spec ConfigMapPropagationSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the EventType.
+	// This data may be out of date.
+	// +optional
+	Status ConfigMapPropagationStatus `json:"status,omitempty"`
+}
+
+var (
+	// Check that ConfigMapPropagation can be validated, can be defaulted, and has immutable fields.
+	_ apis.Validatable = (*ConfigMapPropagation)(nil)
+	_ apis.Defaultable = (*ConfigMapPropagation)(nil)
+
+	// Check that ConfigMapPropagation can return its spec untyped.
+	_ apis.HasSpec = (*ConfigMapPropagation)(nil)
+
+	_ runtime.Object = (*ConfigMapPropagation)(nil)
+
+	// Check that we can create OwnerReferences to a ConfigMapPropagation.
+	_ kmeta.OwnerRefable = (*ConfigMapPropagation)(nil)
+)
+
+type ConfigMapPropagationSpec struct {
+	// OriginalNamespace is the namespace where the original configMaps are in
+	OriginalNamespace string `json:"originalNamespace,omitempty"`
+	// Selector only selects original configMaps with corresponding labels
+	Selector map[string]string `json:"selector,omitempty"`
+}
+
+// ConfigMapPropagationStatus represents the current state of a ConfigMapPropagation.
+type ConfigMapPropagationStatus struct {
+	// inherits duck/v1 Status, which currently provides:
+	// * ObservedGeneration - the 'Generation' of the Service that was last processed by the controller.
+	// * Conditions - the latest available observations of a resource's current state.
+	duckv1.Status `json:",inline"`
```

Comment:
> If I understand the API correctly, you can specify a namespace and a selector which specifies which configmaps are to be propagated?

Yes, it will copy multiple ConfigCaps from a namespace (specified in the `spec.originalnamespace`) with specified selector to current namespace. 

> I think we'd want to know which configmaps were copied and probably some version information (generation?) or something like that.

Currently, there will be a Normal-level Event under this API resource when one ConfigMap copied successfully, showing the name of the original ConfigMap. And there will be a Warn-level Event when one ConfigMap copy failed, showing the original ConfigMap name and the error message. For the copied ConfigMap in current namespace, it will have a ownerreference. This ownerreference points to the configmappropagation which creates this configmap. And the copied configmap’s name follows the rule `configmappropagation name + “-” + original ConfigMap name`” So the ConfigMap information can be checked back implicitly. Those are works have been done related for ConfigMap information. 

I am not sure if those work provide sufficient information...Should I add more component under this API's status struct? (like an array of ObjectReference for each copied ConfigMap)

Reply:
Yeah, so I think the status field should reflect what happened and what didn't happen. In other objects which create composite resources, the state of each "child" resource is reflected in the status field, for example, in the Subscription we don't just say Ready, we reflect dependency statuses in the Status field.
I think as a user, having to dig through the OwnerRefs is not really the most friendly way to figure out what happened.
My preference would be to have an array of each mirrored ConfigMap and maybe even the Generation (maybe overkill?) and that array then turns into "Ready" state if each attempted operation succeeded. I also think that simply saying copy succeeded is not enough, say you start with copying 3 successfully (status is Ready), then one gets deleted from the source, but deletion from the target fails (so you still have 3 though you should only have 2), then I'd argue that this should be a failure, since the mirror operation did not succeed. So, perhaps the array should be:
 - CMPropagation {
      Source: srcNS/scrName
      Op: copy|delete
      Ready: true|false|unknown(?)
      Reason: some string here in error case
}
Or some array like that  so that as a user if the resource does not get Ready, I can figure out what it was trying to do (for example, I don't want to execute the label selector query to see which sources _should_ be even attempted to be copied/removed).
Does that make sense?

Reply:
Yes, this will be much more user-friendly. Thank you for this detailed and informative suggestion! I'll add this status modification.

Reply:
when removing `knative.dev/config-propagation:copy` I get this:

```
NAME               READY   REASON              ORIGINALNAMESPACE   SELECTOR
knative-eventing   False   PropagationFailed   knative-eventing    map[knative.dev/config-category:eventing]
```

Should `ready` be `True` and `Op` be `nocopy` (or something similar). Also `SELECTOR` does not format properly.

Reply:
> Should `ready` be `True` and `Op` be `nocopy` (or something similar). Also `SELECTOR` does not format properly.

I just replied in https://github.com/knative/eventing/issues/1758#issuecomment-574444026 regarding this 😄  I'll modify it for the case when `knative.dev/config-propagation:copy` removed.